### PR TITLE
Replace Razor Source Generator reference with one loaded from a VSIX

### PR DESCRIPTION
**Pull request by tmat**
When analyzers are being added by the project system, replace Razor source generator and its dependencies as they ship in SDK with new source generator and its dependencies from VSIX.

Related Razor change: https://github.com/dotnet/razor-tooling/pull/6821

### CODE REVIEW

**Comment by tmat** [comment]
@jasonmalinowski @chsienki @jaredpar

File: src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs [hunk 1]
```diff
@@ -879,78 +884,107 @@ public void AddAnalyzerReference(string fullPath)
...
+                        {
+                            _workspace.ApplyChangeToWorkspace(w => w.OnAnalyzerReferenceAdded(Id, visualStudioAnalyzer.GetReference()));
+                        }
                     }
                 }
             }
         }
 
-        public void RemoveAnalyzerReference(string fullPath)
+        public void RemoveAnalyzerReference(string fullPathX)
```
Comment (jaredpar) [inline src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs:925]: Why the name change here?

File: src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs [hunk 2]
```diff
@@ -879,78 +884,107 @@ public void AddAnalyzerReference(string fullPath)
...
+                        visualStudioAnalyzer.Dispose();
                     }
                 }
-                else
+            }
+        }
+
+        private IEnumerable<string> GetMapedAnalyzerPaths(string fullPath)
+        {
+            if (fullPath.Split(s_directorySeparator) is [.., "Sdks", "Microsoft.NET.Sdk.Razor", "source-generators", var fileName])
```
Comment (jaredpar) [inline src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs:969]: The dependency on path casing worries me a bit. Essentially if one team changes the casing of anything here then suddenly our solution breaks down. It shouldn't happen ideally but it has happened a few times by accident. We made such a mistake fixing up razor this summer. 

Is there a reason we can't just depend on the file name only vs. the path?

Comment (tmat) [reply src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs:969]: We need to handle dependencies of the generator dll. Replace the ones referenced by the SDK version with those referenced by the VSIX version.

Comment (tmat) [reply src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs:969]: We can make the path comparison case-insensitive.

File: src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs (hunk 1 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs:925]: Accidental.

File: src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs [hunk 3]
```diff
@@ -924,33 +931,63 @@ public void RemoveAnalyzerReference(string fullPath)
...
+                        _workspace.ApplyChangeToWorkspace(w => w.OnAnalyzerReferenceRemoved(Id, visualStudioAnalyzer.GetReference()));
+
+                        visualStudioAnalyzer.Dispose();
                     }
                 }
-                else
+            }
+        }
+
+        private IEnumerable<string> GetMapedAnalyzerPaths(string fullPath)
```
Comment (davidwengier) [inline src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs:967]: Typo:
```suggestion
        private IEnumerable<string> GetMappedAnalyzerPaths(string fullPath)
```

**Review by jaredpar — APPROVED** [review]

**Review by jasonmalinowski — COMMENTED** [review]
Can we add tests for this?

File: src/VisualStudio/Core/Def/Diagnostics/VisualStudioDiagnosticAnalyzerProvider.Factory.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics
+{
+    internal partial class VisualStudioDiagnosticAnalyzerProvider
+    {
+        [Export(typeof(IVisualStudioDiagnosticAnalyzerProviderFactory)), Shared]
+        internal sealed class Factory
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/VisualStudio/Core/Def/Diagnostics/VisualStudioDiagnosticAnalyzerProvider.Factory.cs:19]: Fixed.

File: src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb [hunk 5]
```diff
@@ -108,6 +111,100 @@ Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.ProjectSystemShim
...
+                        Path.Combine(TempRoot.Root, "File.dll")
+                     },
+                     "AnotherExtension")
+                }
+
+                Dim project = Await environment.ProjectFactory.CreateAndAddToWorkspaceAsync(
+                    "Project", LanguageNames.CSharp, CancellationToken.None)
+
+                ' adding just Razor dependency and not the main source generator is a no-op
+                Using project.CreateBatchScope()
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb:136]: You don't need the batch scopes in this case; not hurting anything but would tighten up the test code.

File: src/VisualStudio/Core/Def/Diagnostics/VisualStudioDiagnosticAnalyzerProvider.Factory.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            }
+
+            public async Task<VisualStudioDiagnosticAnalyzerProvider> GetOrCreateProviderAsync(CancellationToken cancellationToken)
+            {
+                // the following code requires UI thread:
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                return GetOrCreateProviderOnMainThread();
+            }
+
+            private VisualStudioDiagnosticAnalyzerProvider GetOrCreateProviderOnMainThread()
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Diagnostics/VisualStudioDiagnosticAnalyzerProvider.Factory.cs:42]: Just inline this into the earlier function at this point?

**Review by jasonmalinowski — APPROVED** [review]

**Pull request merged**